File: web/src/mocks/vulnerabilities.ts
// MSW handlers for the vulnerability endpoints (internal/api/vulnerabilities.go, docs/contracts/api.md
// "Vulnerabilities"): three advisories over the mock fleet, one of them without a fix, plus the catalog
// status that tells an empty list apart from a feed that never synced.
import { http, HttpResponse, type HttpResponseResolver } from "msw";
import type { VulnFinding, VulnGroup } from "@/api/vulnerabilities";
import { authenticate } from "./account";
import { formatTs, HOST_IDS } from "./fixtures";

const API = "*/api/v1";

function authed(resolver: HttpResponseResolver): HttpResponseResolver {
  return (info) => {
    const ctx = authenticate(info.request);
    if (ctx instanceof Response) return ctx;
    return resolver(info);
  };
}

const now = Date.now();

const FINDINGS: VulnFinding[] = [
  {
    host_id: HOST_IDS.web,
    host_name: "web-1.shop.internal",
    vuln_id: "DSA-5600-1",
    cve: "CVE-2026-0001",
    severity: "critical",
    score: 9.8,
    ecosystem: "Debian:12",
    package: "libexample",
    version: "1.2.3-4+deb12u1",
    fixed_in: "1.2.3-4+deb12u2",
    summary: "buffer overflow in libexample",
    first_seen: formatTs(now - 6 * 86_400_000),
    last_seen: formatTs(now - 3_600_000),
  },
  {
    host_id: HOST_IDS.db,
    host_name: "db-1.shop.internal",
    vuln_id: "DSA-5600-1",
    cve: "CVE-2026-0001",
    severity: "critical",
    score: 9.8,
    ecosystem: "Debian:12",
    package: "libexample",
    version: "1.2.3-4+deb12u1",
    fixed_in: "1.2.3-4+deb12u2",
    summary: "buffer overflow in libexample",
    first_seen: formatTs(now - 6 * 86_400_000),
    last_seen: formatTs(now - 3_600_000),
  },
  {
    host_id: HOST_IDS.web,
    host_name: "web-1.shop.internal",
    vuln_id: "CVE-2026-0042",
    cve: "CVE-2026-0042",
    severity: "high",
    score: 7.5,
    ecosystem: "Debian:12",
    package: "openssl",
    version: "3.0.11-1~deb12u2",
    // No fixed version yet: the case a person cannot resolve by upgrading.
    fixed_in: "",
    summary: "denial of service in the TLS handshake",
    first_seen: formatTs(now - 2 * 86_400_000),
    last_seen: formatTs(now - 3_600_000),
  },
  {
    host_id: HOST_IDS.web,
    host_name: "web-1.shop.internal",
    vuln_id: "CVE-2026-0100",
    cve: "CVE-2026-0100",
    severity: "medium",
    score: 5.3,
    ecosystem: "Debian:12",
    package: "curl",
    version: "7.88.1-10+deb12u4",
    fixed_in: "7.88.1-10+deb12u5",
    summary: "information disclosure in cookie handling",
    first_seen: formatTs(now - 86_400_000),
    last_seen: formatTs(now - 3_600_000),
  },
];

function groups(severity?: string): VulnGroup[] {
  const byVuln = new Map<string, VulnGroup>();
  for (const f of FINDINGS) {
    if (severity && f.severity !== severity) continue;
    const g = byVuln.get(f.vuln_id);
    if (g) {
      g.hosts += 1;
      if (!g.packages.includes(f.package)) g.packages.push(f.package);
      continue;
    }
    byVuln.set(f.vuln_id, {
      vuln_id: f.vuln_id,
      cve: f.cve,
      severity: f.severity,
      score: f.score,
      summary: f.summary,
      hosts: 1,
      packages: [f.package],
      first_seen: f.first_seen,
      last_seen: f.last_seen,
    });
  }
  return [...byVuln.values()];
}

export const vulnerabilityHandlers = [
  http.get(`${API}/vulnerabilities`, authed(({ request }) => {
    const url = new URL(request.url);
    const severity = url.searchParams.get("severity") ?? undefined;
    const list = groups(severity);
    const counts: Record<string, number> = { critical: 0, high: 0, medium: 0, low: 0, none: 0 };
    for (const g of groups()) counts[g.severity] = (counts[g.severity] ?? 0) + 1;
    return HttpResponse.json({ vulnerabilities: list, severity_counts: counts });
  })),
  http.get(`${API}/vulnerabilities/catalog/status`, authed(() =>
    HttpResponse.json({
      vulnerabilities: 61_234,
      affected_ranges: 98_211,
      ecosystems: 2,
      last_synced_at: formatTs(now - 4 * 3_600_000),
      sources: [
        { source: "osv", ecosystem: "Debian:12", synced_at: formatTs(now - 4 * 3_600_000), count: 41_022, error: "" },
        { source: "osv", ecosystem: "Ubuntu:22.04", synced_at: formatTs(now - 4 * 3_600_000), count: 20_212, error: "" },
      ],
    }),
  )),
  http.get(`${API}/vulnerabilities/:id`, authed(({ params }) => {
    const findings = FINDINGS.filter((f) => f.vuln_id === params.id);
    if (findings.length === 0) {
      return HttpResponse.json({ error: { code: "not_found", message: "no open finding for this advisory" } }, { status: 404 });
    }
    return HttpResponse.json({ vuln_id: params.id, findings });
  })),
  http.get(`${API}/hosts/:hostId/vulnerabilities`, authed(({ params }) =>
    HttpResponse.json({ host_id: params.hostId, findings: FINDINGS.filter((f) => f.host_id === params.hostId) }),
  )),
];

File: internal/api/vulnerabilities.go
package api

import (
	"net/http"
	"strconv"
	"strings"

	"github.com/onuragtas/openlog/internal/api/query"
	"github.com/onuragtas/openlog/internal/vuln"
)

// Vulnerabilities (docs/contracts/api.md "Vulnerabilities", D-142). The findings are per host and therefore
// tenant data, read through the tenant-scoped query layer like every other telemetry read; the advisories
// they name come from a catalog that is shared by the whole installation, because OSV is public data.
//
// Reads only: a finding is produced by the matcher, not by a person. What a person does about it happens on
// the host.

// maxVulnRows bounds a listing.
const maxVulnRows = 1000

// SetVulnerabilities enables /api/v1/vulnerabilities/* with the catalog behind them. Must be called before Run.
func (s *Server) SetVulnerabilities(catalog vuln.Catalog) {
	s.vulnCatalog = catalog
	s.srv.Handler = s.Handler()
}

func (s *Server) vulnRoutes(mux *http.ServeMux) {
	route := func(pattern string, h handlerFunc) { mux.Handle(pattern, s.wrap(pattern, h)) }
	route("GET /api/v1/vulnerabilities", s.listVulnerabilities)
	route("GET /api/v1/vulnerabilities/{id}", s.getVulnerability)
	route("GET /api/v1/hosts/{host_id}/vulnerabilities", s.hostVulnerabilities)
	if s.vulnCatalog != nil {
		route("GET /api/v1/vulnerabilities/catalog/status", s.vulnerabilityCatalogStatus)
	}
}

type vulnGroupJSON struct {
	VulnID    string   `json:"vuln_id"`
	CVE       string   `json:"cve"`
	Severity  string   `json:"severity"`
	Score     float64  `json:"score"`
	Summary   string   `json:"summary"`
	Hosts     uint64   `json:"hosts"`
	Packages  []string `json:"packages"`
	FirstSeen string   `json:"first_seen"`
	LastSeen  string   `json:"last_seen"`
}

type vulnFindingJSON struct {
	HostID    string  `json:"host_id"`
	HostName  string  `json:"host_name"`
	VulnID    string  `json:"vuln_id"`
	CVE       string  `json:"cve"`
	Severity  string  `json:"severity"`
	Score     float64 `json:"score"`
	Ecosystem string  `json:"ecosystem"`
	Package   string  `json:"package"`
	Version   string  `json:"version"`
	FixedIn   string  `json:"fixed_in"`
	Summary   string  `json:"summary"`
	FirstSeen string  `json:"first_seen"`
	LastSeen  string  `json:"last_seen"`
}

// listVulnerabilities groups the open findings by advisory: the question a person starts with is "what is
// wrong with my fleet", not "what is wrong with host 7".
func (s *Server) listVulnerabilities(w http.ResponseWriter, r *http.Request, sc *query.Scope) error {
	noStore(w)
	q := sc.From(query.HostVulnerabilities).Final().Columns(
		"vuln_id AS id",
		"any(cve) AS cve_id",
		"any(severity) AS sev",
		"toFloat64(max(score)) AS sc",
		"any(summary) AS sum",
		// Not `AS hosts`: the query layer refuses any fragment containing a table name, and `hosts` is one.
		"uniqExact(host_id) AS host_count",
		"arraySlice(arraySort(groupUniqArray(package)), 1, 20) AS pkgs",
		"toInt64(toUnixTimestamp64Milli(min(first_seen))) AS first_ms",
		"toInt64(toUnixTimestamp64Milli(max(last_seen))) AS last_ms",
	).Where("resolved_at = toDateTime(0)").GroupBy("vuln_id").Limit(maxVulnRows)
	if sev := r.URL.Query().Get("severity"); sev != "" {
		if !knownSeverity(sev) {
			return &apiError{http.StatusBadRequest, "invalid_argument", "severity must be one of " + strings.Join(vuln.Severities, ", ")}
		}
		q.Where("severity = {severity:String}").Param("severity", sev)
	}
	if hostID := r.URL.Query().Get("host_id"); hostID != "" {
		q.Where("host_id = {host_id:String}").Param("host_id", hostID)
	}
	// Critical first, then the advisory that affects the most hosts: the order a person works down.
	q.OrderBy("multiIf(sev = 'critical', 0, sev = 'high', 1, sev = 'medium', 2, sev = 'low', 3, 4)", "host_count DESC", "id")

	rows, err := sc.Query(r.Context(), q)
	if err != nil {
		return err
	}
	defer rows.Close()
	out := []vulnGroupJSON{}
	counts := map[string]uint64{}
	for rows.Next() {
		var (
			g               vulnGroupJSON
			pkgs            []string
			firstMs, lastMs int64
			hosts           uint64
		)
		if err := rows.Scan(&g.VulnID, &g.CVE, &g.Severity, &g.Score, &g.Summary, &hosts, &pkgs, &firstMs, &lastMs); err != nil {
			return err
		}
		g.Hosts, g.Packages = hosts, pkgs
		if g.Packages == nil {
			g.Packages = []string{}
		}
		g.FirstSeen, g.LastSeen = formatMillis(firstMs), formatMillis(lastMs)
		counts[g.Severity]++
		out = append(out, g)
	}
	if err := rows.Err(); err != nil {
		return err
	}
	severities := map[string]uint64{}
	for _, sev := range vuln.Severities {
		severities[sev] = counts[sev]
	}
	writeJSON(w, http.StatusOK, map[string]any{"vulnerabilities": out, "severity_counts": severities})
	return nil
}

// getVulnerability lists the hosts one advisory was found on.
func (s *Server) getVulnerability(w http.ResponseWriter, r *http.Request, sc *query.Scope) error {
	noStore(w)
	id := r.PathValue("id")
	if id == "" {
		return &apiError{http.StatusBadRequest, "invalid_argument", "the advisory id is required"}
	}
	findings, err := s.vulnFindings(r, sc, "vuln_id = {vuln_id:String}", "vuln_id", id)
	if err != nil {
		return err
	}
	if len(findings) == 0 {
		return notFound("no open finding for this advisory")
	}
	writeJSON(w, http.StatusOK, map[string]any{"vuln_id": id, "findings": findings})
	return nil
}

// hostVulnerabilities lists the open findings of one host, which is what the host's own screen shows.
func (s *Server) hostVulnerabilities(w http.ResponseWriter, r *http.Request, sc *query.Scope) error {
	noStore(w)
	hostID := r.PathValue("host_id")
	if err := requireHost(r, sc, hostID); err != nil {
		return err
	}
	findings, err := s.vulnFindings(r, sc, "host_id = {host_id:String}", "host_id", hostID)
	if err != nil {
		return err
	}
	writeJSON(w, http.StatusOK, map[string]any{"host_id": hostID, "findings": findings})
	return nil
}

func (s *Server) vulnFindings(r *http.Request, sc *query.Scope, where, param, value string) ([]vulnFindingJSON, error) {
	limit := maxVulnRows
	if v := r.URL.Query().Get("limit"); v != "" {
		n, err := strconv.Atoi(v)
		if err != nil || n < 1 || n > maxVulnRows {
			return nil, &apiError{http.StatusBadRequest, "invalid_argument", "limit must be between 1 and 1000"}
		}
		limit = n
	}
	q := sc.From(query.HostVulnerabilities).Final().Columns(
		"host_id AS hid", "host_name AS hname", "vuln_id AS vid", "cve AS cve_id", "severity AS sev",
		"toFloat64(score) AS sc", "ecosystem AS eco", "package AS pkg", "version AS ver",
		"fixed_in AS fixed", "summary AS sum",
		"toInt64(toUnixTimestamp64Milli(first_seen)) AS first_ms",
		"toInt64(toUnixTimestamp64Milli(last_seen)) AS last_ms",
	).Where("resolved_at = toDateTime(0)").Where(where).Param(param, value).
		OrderBy("multiIf(sev = 'critical', 0, sev = 'high', 1, sev = 'medium', 2, sev = 'low', 3, 4)", "pkg", "vid").
		Limit(limit)
	rows, err := sc.Query(r.Context(), q)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := []vulnFindingJSON{}
	for rows.Next() {
		var f vulnFindingJSON
		var firstMs, lastMs int64
		if err := rows.Scan(&f.HostID, &f.HostName, &f.VulnID, &f.CVE, &f.Severity, &f.Score, &f.Ecosystem,
			&f.Package, &f.Version, &f.FixedIn, &f.Summary, &firstMs, &lastMs); err != nil {
			return nil, err
		}
		f.FirstSeen, f.LastSeen = formatMillis(firstMs), formatMillis(lastMs)
		out = append(out, f)
	}
	return out, rows.Err()
}

// vulnerabilityCatalogStatus says how fresh the catalog is, so "no findings" can be told from "the feed
// never synced" — the difference between a clean fleet and a broken feature.
func (s *Server) vulnerabilityCatalogStatus(w http.ResponseWriter, r *http.Request, _ *query.Scope) error {
	noStore(w)
	stats, err := s.vulnCatalog.Stats(r.Context())
	if err != nil {
		return err
	}
	syncs, err := s.vulnCatalog.SyncStatus(r.Context())
	if err != nil {
		return err
	}
	out := make([]map[string]any, 0, len(syncs))
	for _, st := range syncs {
		out = append(out, map[string]any{"source": st.Source, "ecosystem": st.Ecosystem,
			"synced_at": formatTime(st.SyncedAt), "count": st.Count, "error": st.Error})
	}
	body := map[string]any{"vulnerabilities": stats.Vulnerabilities, "affected_ranges": stats.AffectedRanges,
		"ecosystems": stats.Ecosystems, "last_synced_at": nil, "sources": out}
	if stats.LastSyncedAt != nil {
		body["last_synced_at"] = formatTime(*stats.LastSyncedAt)
	}
	writeJSON(w, http.StatusOK, body)
	return nil
}

func knownSeverity(s string) bool {
	for _, v := range vuln.Severities {
		if v == s {
			return true
		}
	}
	return false
}
